source/common/router/router.cc

```diff
@@ -99,7 +99,7 @@ void FilterUtility::setUpstreamScheme(Http::RequestHeaderMap& headers, bool down
 
 bool FilterUtility::shouldShadow(const ShadowPolicy& policy, Runtime::Loader& runtime,
                                  uint64_t stable_random) {
-  if (policy.cluster().empty()) {
+  if (policy.cluster().empty() && policy.clusterHeader().get().empty()) {
```

Comment:
seems there is no chance to be empty for both, it should be an assertion here.

Reply:
I think there should be an assertion that not both are set

Reply:
Thank you all! Done, added assertion here.